c2rust-analyze/src/borrowck/mod.rs

```diff
@@ -15,25 +20,136 @@ mod dump;
 mod type_check;
 
 #[derive(Clone, Copy, PartialEq, Eq, Debug, Hash, Default)]
-pub struct Label {
+pub struct Label<'tcx> {
+    /// The [`Origin`] of this type
     pub origin: Option<Origin>,
+    /// The [Origins](`Origin`) associated with each lifetime
+    /// parameter of this type, if applicable
+    pub origin_params: Option<&'tcx [(OriginParam, Origin)]>,
     pub perm: PermissionSet,
 }
 
-pub type LTy<'tcx> = LabeledTy<'tcx, Label>;
-pub type LTyCtxt<'tcx> = LabeledTyCtxt<'tcx, Label>;
+pub type LTy<'tcx> = LabeledTy<'tcx, Label<'tcx>>;
+pub type LTyCtxt<'tcx> = LabeledTyCtxt<'tcx, Label<'tcx>>;
+
+/// Metadata describing lifetimes and lifetime parameters
+/// of a struct field.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FieldMetadata<'tcx> {
+    /// The lifetime of the field, e.g. `*mut &'a mut foo_type`
+    /// would have an index set of {'h0, 'a}
+    pub lifetime: IndexSet<OriginArg<'tcx>>,
+    /// The lifetime parameters of a field, e.g. if a struct
+    /// `foo<'a, 'b>` is a field of `bar<'c, 'd>` as field: `foo<'c, 'd>`,
+    /// the lifetime params would be a set {'c, 'd}
+    pub lifetime_params: IndexSet<OriginArg<'tcx>>,
```

Comment:
If I'm understanding this correctly, then a  type like `*mut &'a Foo<'b>` (where `Foo` contains a pointer and therefore has a hypothetical lifetime parameter) would have metadata like this:
```
lifetime = ['h0, 'a]
lifetime_params = ['b, 'h1]
```
But this representation only works when the field type has a specific shape (some number of pointers/refs, with the final pointee type possibly being an ADT).  Would it be reasonable to instead use something like `LabeledTy<&[OriginArg]>`, where each type is labeled with a list of `OriginArg`s, similar to the `origin_params` field of `Label`?  (For ref/ptr, you could use a list of length one.)  With this representation, we could handle (to some extent) arbitrary types appearing in partially-refactored code, like `field1: (*mut i32, *mut i32)`  or `field2: Vec<MyAdt>`.

Reply:
I think that's reasonable and a good idea. I've put some changes in 1e9a0fb6. Would you mind taking a look?

Reply:
It sounds like this handles `*mut (ADT, ADT)` but not `(*mut ADT, *mut ADT)` - is that correct?

In general, I think we should avoid making unnecessary distinctions between `Ref`/`RawPtr` and `Adt` types.  Given `struct MyRef<'a>(&'a i32)`, fields `x: &'b i32` and `y: MyRef<'b>` should be treated almost identically in most of the field handling code.  In particular, I don't think it's necessary to have this split between `lifetime` (`OriginArg`s for outer `Ref`/`RawPtr` types) and `lifetime_params` (`OriginArg`s for everything else).  `&'b i32` and `MyRef<'b>` each have a single `OriginArg::Actual`, which gets substituted the same way when the field is accessed - why does one have its `OriginArg` stored in `lifetime` while the other's goes in `lifetime_params`?

Specifically, I think both `lifetime` and `lifetime_params` can be combined into a single `LabeledTy` field like the one you added in that commit.  This should allow ref/ptr and ADT types to be handled more uniformly.

Reply:
Yes, you're right. I've made the changes in 9f5ad68b

Reply:
also added test case in fae34e90